numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -3692,9 +3691,12 @@ static void
         }
     }
     else {
-        for (i = 0; i < ni; i++) {
+        for (i = 0, j = 0; i < ni; i++, j++) {
+            if (j >= nv) {
+                j = 0;
+            }
```

Comment:
Speaking of branch prediction and code bumming, `gcc -O2` generates a loop with three conditional jumps, but if this conditional block is replaced with

```c
j *= (j < nv);
```

there's only two of them (at the expense of a multiplication). Might be worth a try.